src/main/scala/devices/debug/Debug.scala

```diff
@@ -289,11 +297,110 @@ class TLDebugModuleOuter(device: Device)(implicit p: Parameters) extends LazyMod
-      0 -> Seq(RWNotify(32, DMCONTROLRdData.asUInt(),
-        DMCONTROLWrDataVal, DMCONTROLRdEn, DMCONTROLWrEn, Some(RegFieldDesc("dmi_dmcontrol", "", reset=Some(0)))))
-    )
+    //--------------------------------------------------------------
+    // Hart array mask and window
+    //  HAMASK is 1 bit per component
+    //  HAWINDOWSEL selects a 32-bit slice of HAMASK to be visible for read/write in HAWINDOW
+    //--------------------------------------------------------------
+    val hamask = Wire(init = Vec.fill(nComponents){false.B})
+
+      // The following need to be declared even if supportHartArray is false due to reference
+      // at compile time by dmiNode.regmap
+    val HAWINDOWSELRdData = Wire(init = (new HAWINDOWSELFields()).fromBits(0.U))
+    val HAWINDOWSELWrDataVal = Wire(init = 0.U(32.W))
+    val HAWINDOWSELWrData = (new HAWINDOWSELFields()).fromBits(HAWINDOWSELWrDataVal)
+    val HAWINDOWSELWrEn   = Wire(init = false.B)
+    val HAWINDOWSELRdEn   = Wire(init = false.B)
+
+    val HAWINDOWRdData = Wire(init = (new HAWINDOWFields()).fromBits(0.U))
+    val HAWINDOWWrDataVal = Wire(init = 0.U(32.W))
+    val HAWINDOWWrData = (new HAWINDOWFields()).fromBits(HAWINDOWWrDataVal)
+    val HAWINDOWWrEn   = Wire(init = false.B)
+    val HAWINDOWRdEn   = Wire(init = false.B)
+
+    if (supportHartArray) {
+      val HAWINDOWSELReset = Wire(init = (new HAWINDOWSELFields().fromBits(0.U)))
+      val HAWINDOWSELNxt = Wire(init = (new HAWINDOWSELFields().fromBits(0.U)))
+      val HAWINDOWSELReg = Wire(init = new HAWINDOWSELFields().fromBits(AsyncResetReg(updateData = HAWINDOWSELNxt.asUInt,
+        resetData = 0,
+        enable = true.B,
+        name = "HAWINDOWSEL"
+      )))
+
+      HAWINDOWSELRdData := Wire(init = HAWINDOWSELReg)
+      HAWINDOWSELNxt := HAWINDOWSELReg
+      when (~dmactive | ~supportHartArray.B) {
+        HAWINDOWSELNxt := HAWINDOWSELReset
+      } .otherwise {
+        when (HAWINDOWSELWrEn) {
+            // Unneeded upper bits of HAWINDOWSEL are tied to 0.  Entire register is 0 if 32 or fewer harts.
+          if (nComponents > 32) {
+            HAWINDOWSELNxt.hawindowsel := HAWINDOWSELWrData.hawindowsel & (log2Up(nComponents) - 5).U
+          } else {
+            HAWINDOWSELNxt.hawindowsel := 0.U
+          }
+        }
+      }
+
+      val numHAMASKSlices = ((nComponents - 1)/32)+1
+      HAWINDOWRdData.maskdata := 0.U     // default, overridden below
+      for (ii <- 0 until numHAMASKSlices) {
+        val SliceMask = if (nComponents > ii*32 + 31) 0xFFFFFFFF  // All harts in this slice exist
+                        else (1<<(nComponents - ii*32)) - 1       // Partial last slice
+        val HAMASKRst = Wire(init = (new HAWINDOWFields().fromBits(0.U)))
+        val HAMASKNxt = Wire(init = (new HAWINDOWFields().fromBits(0.U)))
+        val HAMASKReg = Wire(init = Vec(AsyncResetReg(updateData = HAMASKNxt.asUInt,
+          resetData = 0,
+          enable = true.B,
+          name = "HAMASK")))
```

Comment:
this will make a bunch of registers called "HAMASK" which then will rely on Chisel/FIRRTL to give them different names. Can you add `ii` to the name, ie `name = s"HMASKReg${ii}"`